src/puppetlabs/puppetdb/cli/benchmark.clj

```diff
@@ -472,27 +429,148 @@
+  (let [random-entity (fn [host entities]
+                        (some-> entities
+                                rand-nth
+                                (assoc "certname" host)))]
+    (for [i (range n)]
+      (let [host (str "host-" i)]
+        {:host host
+         :catalog (when-let [cat (random-entity host catalogs)]
+                    (update cat "resources"
+                            (partial map #(update % "tags"
+                                                  conj
+                                                  pdb-host))))
+         :report (random-entity host reports)
+         :factset (random-entity host facts)}))))
+
+(defn populate-queue
+  "Fills queue with host entries.  Stops if mq is closed."
+  [mq numhosts run-interval pdb-host catalogs reports facts]
+  (println-err (trs "Populating queue in the background"))
+  (doseq [host (random-hosts numhosts pdb-host catalogs reports facts)
+          :while (>!! mq (assoc host :lastrun (rand-lastrun run-interval)))]
+    true)
+  (println-err (trs "Finished populating queue")))
+
+(defn cycle-commands [numhosts run-interval rand-perc simulation-threads
+                      rate-monitor-ch command-send-ch
+                      pdb-host catalogs reports facts]
+  "Populates the message queue with messages, and then cycles through
+  those messages, sending and altering each one every time it
+  reappears at the head of the queue.  Returns a stop function of no
+  arguments that can be used to effect an orderly shutdown."
+  ;; The stop function is also registered as a shutdown hook (thread),
+  ;; and is set up as a oneshot, protected by an atomic boolean.  All
+  ;; of the work that needs to be completed on a SIGINT, etc. is
+  ;; coordinated from that hook via the stop function because after a
+  ;; signal, only daemon threads, shutdown hooks, and finalizers will
+  ;; still run, and daemon threads will only run until all of the
+  ;; hooks and finalizers are finished.
+  (let [preftmp (get-path (or (System/getenv "TMPDIR")
+                              (System/getProperty "java.io.tmpdir")))
+        tmpdir (Files/createTempDirectory preftmp
+                                          "pdb-bench-"
+                                          (into-array FileAttribute []))
+        _ (println-err (trs "Using scratch dir {0}" (pr-str (str tmpdir))))
+        message-dir (doto (.resolve tmpdir "q") fs/mkdir)
+        mq (chan (message-buffer message-dir numhosts))
+        rate-limited-mq-out (chan)
+        channels [rate-limited-mq-out mq rate-monitor-ch command-send-ch]
+        fill-queue (doto (Thread. #(populate-queue mq numhosts run-interval
+                                                   pdb-host
+                                                   catalogs reports facts))
+                     (.setDaemon true))
+        shutdown-thread (promise)  ;; Because no letrec...
+        stopping? (AtomicBoolean.)
+        stop (fn []
```

Comment:
I've spent some time understanding this and I think I see a way to simplify it. I'll chat with you about it when you're around.